File: dart/lib/lexer.dart
import 'package:dart/token.dart';

class Lexer {
  String input;
  int position = 0;
  int readPosition = 0;
  String ch = '';
  Lexer({required this.input}) {
    readChar();
  }

  @override
  String toString() {
    return "($position, $readPosition, $ch)";
  }

  void readChar() {
    readPosition >= input.length ? ch = '' : ch = input[readPosition];
    position = readPosition;
    readPosition += 1;
  }

  String peekChar() {
    return readPosition >= input.length ? '' : input[readPosition];
  }

  bool isLetter(String ch) {
    return 'a'.codeUnitAt(0) <= ch.codeUnitAt(0) &&
            ch.codeUnitAt(0) <= 'z'.codeUnitAt(0) ||
        'A'.codeUnitAt(0) <= ch.codeUnitAt(0) &&
            ch.codeUnitAt(0) <= 'Z'.codeUnitAt(0) ||
        ch == '_';
  }

  bool isDigit(String ch) {
    return switch (int.tryParse(ch)) {
      null => false,
      _ => true,
    };
  }

  String readIdentifier() {
    var pos = position;
    while (isLetter(ch)) {
      readChar();
    }
    return input.substring(pos, position);
  }

  String readNumber() {
    var pos = position;
    while (isDigit(ch)) {
      readChar();
    }
    return input.substring(pos, position);
  }

  void skipWhiteSpace() {
    while (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r') {
      readChar();
    }
  }

  Token checkIdentOrKeyword() {
    var literal = readIdentifier();
    var type = lookupIdent(literal);
    return Token(type: type, literal: literal);
  }

  Token getNumber() {
    var type = TokenType.INT;
    var literal = readNumber();
    return Token(type: type, literal: literal);
  }

  Token getString() {
    var pos = position;
    // Skip the first '
    readChar();
    // Read the entire string within ''
    while (ch != '\'' && ch != '"') {
      readChar();
    }
    // Skip the last '
    readChar();
    var literal = input.substring(pos, position);
    return Token(type: TokenType.STRING, literal: literal);
  }

  Token createToken(TokenType type, {String literal = ''}) {
    readChar();
    return Token(type: type, literal: literal == '' ? type.id : literal);
  }

  Token handleTwoCharToken() {
    var literal = ch;
    if (peekChar() == '=') {
      readChar();
      literal += ch;
    }
    return switch (literal) {
      '==' => createToken(TokenType.EQ),
      '!=' => createToken(TokenType.NOT_EQ),
      '+=' => createToken(TokenType.PLUS_EQ),
      '-=' => createToken(TokenType.MINUS_EQ),
      '=' => createToken(TokenType.ASSIGN),
      '+' => createToken(TokenType.PLUS),
      '-' => createToken(TokenType.MINUS),
      '!' => createToken(TokenType.BANG),
      '/' => createToken(TokenType.SLASH),
      '*' => createToken(TokenType.ASTERISK),
      '<' => createToken(TokenType.LT),
      '>' => createToken(TokenType.GT),
      _ => createToken(TokenType.ILLEGAL, literal: ch),
    };
  }

  Token handleString(String ch) {
    return !(isLetter(peekChar()) || isDigit(peekChar())) &&
            (peekChar() != '\'' && peekChar() != '"')
        ? (ch == '\''
            ? createToken(TokenType.SQUOTE)
            : createToken(TokenType.DQUOTE))
        : getString();
  }

  Token nextToken() {
    skipWhiteSpace();
    return switch (ch) {
      '=' ||
      '!' ||
      '+' ||
      '-' ||
      '/' ||
      '*' ||
      '<' ||
      '>' =>
        handleTwoCharToken(),
      '\'' || '"' => handleString(ch),
      '(' => createToken(TokenType.LPAREN),
      ')' => createToken(TokenType.RPAREN),
      '{' => createToken(TokenType.LBRACE),
      '}' => createToken(TokenType.RBRACE),
      ',' => createToken(TokenType.COMMA),
      ';' => createToken(TokenType.SEMICOLON),
      '' => createToken(TokenType.EOF),
      _ => isLetter(ch)
          ? checkIdentOrKeyword()
          : isDigit(ch)
              ? getNumber()
              : createToken(TokenType.ILLEGAL, literal: ch),
    };
  }
}

File: rust/src/lexer.rs
use crate::token::Token;

pub struct Lexer {
    input: String,
    pos: usize,
    read_pos: usize,
    ch: char,
}

impl Lexer {
    pub fn new(input: String) -> Self {
        let mut l = Lexer {
            input,
            pos: 0,
            read_pos: 0,
            ch: ' ',
        };
        l.read_char();
        l
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();
        let t = match self.ch {
            '=' | '!' => return self.read_two_char_token(),
            '+' => Token::Plus(self.ch.to_string()),
            '-' => Token::Minus(self.ch.to_string()),
            '*' => Token::Asterisk(self.ch.to_string()),
            '/' => Token::Slash(self.ch.to_string()),
            '<' => Token::LT(self.ch.to_string()),
            '>' => Token::GT(self.ch.to_string()),
            '(' => Token::LParen(self.ch.to_string()),
            ')' => Token::RParen(self.ch.to_string()),
            '{' => Token::LBrace(self.ch.to_string()),
            '}' => Token::RBrace(self.ch.to_string()),
            ',' => Token::Comma(self.ch.to_string()),
            ';' => Token::Semicolon(self.ch.to_string()),
            '\0' => Token::Eof("".to_string()),
            _ => match (self.ch.is_alphabetic(), self.ch.is_numeric()) {
                (true, false) => return self.read_identifier(),
                (false, true) => return self.read_number(),
                _ => Token::Illegal(self.ch.to_string()),
            },
        };
        self.read_char();

        return t;
    }

    fn read_two_char_token(&mut self) -> Token {
        let c = self.peek_char();
        let t = match (self.ch, c) {
            ('=', '=') => {self.read_char(); Token::Eq("==".to_string())},
            ('!', '=') => {self.read_char(); Token::NotEq("!=".to_string())},
            ('=', _) => Token::Assign(self.ch.to_string()),
            ('!', _) => Token::Bang(self.ch.to_string()),
            _ => Token::Illegal(c.to_string())
        };
        self.read_char();
        return t
    }

    fn read_identifier(&mut self) -> Token {
        let pos = self.pos;
        while self.ch.is_alphabetic() {
            self.read_char();
        }
        let s = self.input[pos..self.pos].to_string();
        Token::lookup_ident(&s)
    }

    fn read_number(&mut self) -> Token {
        let pos = self.pos;
        while self.ch.is_numeric() {
            self.read_char();
        }
        let s = self.input[pos..self.pos].to_string();
        Token::Int(s)
    }

    fn read_char(&mut self) {
        if self.read_pos >= self.input.len() {
            self.ch = '\0';
        } else {
            self.ch = self.input.as_bytes()[self.read_pos] as char;
        }
        self.pos = self.read_pos;
        self.read_pos += 1;
    }

    fn peek_char(&mut self) -> char {
        if self.read_pos >= self.input.len() {
            return '\0';
        } else {
            return self.input.as_bytes()[self.read_pos] as char;
        }
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_whitespace() {
            self.read_char();
        }
    }
}
